include/exec/system_scheduler.hpp

```diff
@@ -0,0 +1,685 @@
+  virtual __exec_system_sender_interface* schedule() = 0;
+  // TODO: Move chaining in here to support chaining after a system_sender or other system_bulk_sender
+  // or don't do anything that specific?
+  virtual __exec_system_sender_interface*
+    bulk(__exec_system_bulk_shape __shp, __exec_system_bulk_function_object __fn) = 0;
+  virtual bool equals(const __exec_system_scheduler_interface* __rhs) const = 0;
+};
+
+struct __exec_system_operation_state_interface {
+  virtual void start() noexcept = 0;
+};
+
+struct __exec_system_receiver {
+  void* __cpp_recv_ = nullptr;
+  void (*set_value)(void* __cpp_recv) noexcept;
+  void (*set_stopped)(void* __cpp_recv) noexcept;
+  // Type-erase the exception pointer for extern-c-ness
+  void (*set_error)(void* __cpp_recv, void* __exception) noexcept;
+};
+
+struct __exec_system_sender_interface {
+  virtual __exec_system_operation_state_interface*
+    connect(__exec_system_receiver __recv) noexcept = 0;
+  virtual __exec_system_scheduler_interface* get_completion_scheduler() noexcept = 0;
+};
+
+// Low-level APIs
+// Phase 2 will move to pointers and ref counting ala COM
+// Phase 3 will move these to weak symbols and allow replacement in tests
+// Default implementation based on static_thread_pool
+struct __exec_system_context_impl : public __exec_system_context_interface {
+  exec::static_thread_pool __pool_;
+
+  __exec_system_scheduler_interface* get_scheduler() noexcept override;
+};
+
+struct __exec_system_scheduler_impl : public __exec_system_scheduler_interface {
+  __exec_system_scheduler_impl(
+    __exec_system_context_impl* __ctx,
+    decltype(__ctx->__pool_.get_scheduler()) __pool_scheduler)
+    : __ctx_{__ctx}
+    , __pool_scheduler_{__pool_scheduler} {
+  }
+
+  __exec_system_context_impl* __ctx_;
+  decltype(__ctx_->__pool_.get_scheduler()) __pool_scheduler_;
+
+  __exec_system_sender_interface* schedule() override;
+
+  __exec_system_sender_interface*
+    bulk(__exec_system_bulk_shape __shp, __exec_system_bulk_function_object __fn) override;
+
+  stdexec::forward_progress_guarantee get_forward_progress_guarantee() const override {
+    return stdexec::forward_progress_guarantee::parallel;
+  }
+
+  bool equals(const __exec_system_scheduler_interface* __rhs) const override {
+    auto __rhs_impl = dynamic_cast<const __exec_system_scheduler_impl*>(__rhs);
+    return __rhs_impl && __rhs_impl->__ctx_ == __ctx_;
+  }
+};
+
+struct __exec_system_operation_state_impl;
+using __exec_pool_sender_t =
+  decltype(stdexec::schedule(std::declval<__exec_system_scheduler_impl>().__pool_scheduler_));
+
+struct __exec_system_pool_receiver {
+  using receiver_concept = stdexec::receiver_t;
+
+  friend void tag_invoke(stdexec::set_value_t, __exec_system_pool_receiver&&) noexcept;
+
+  friend void tag_invoke(stdexec::set_stopped_t, __exec_system_pool_receiver&&) noexcept;
+
+  friend void
+    tag_invoke(stdexec::set_error_t, __exec_system_pool_receiver&&, std::exception_ptr) noexcept;
```

Comment:
These should be defined in situ to remain hidden friends.

```suggestion
  void set_value() noexcept;
  void set_stopped() noexcept;
  void set_error(std::exception_ptr) noexcept;

  friend void tag_invoke(stdexec::set_value_t, __exec_system_pool_receiver&& __self) noexcept {
    __self.set_value();
  }

  friend void tag_invoke(stdexec::set_stopped_t, __exec_system_pool_receiver&& __self) noexcept {
    __self.set_stopped();
  }

  friend void
    tag_invoke(stdexec::set_error_t, __exec_system_pool_receiver&& __self, std::exception_ptr __ptr) noexcept {
      __self.set_error(std::move(__ptr));
    }
```